src/com/amazon/ionelement/impl/MutableStructFieldsImpl.kt

```diff
@@ -0,0 +1,109 @@
+package com.amazon.ionelement.impl
+
+import com.amazon.ionelement.api.AnyElement
+import com.amazon.ionelement.api.IonElement
+import com.amazon.ionelement.api.MetaContainer
+import com.amazon.ionelement.api.MutableStructFields
+import com.amazon.ionelement.api.StructElement
+import com.amazon.ionelement.api.StructField
+import com.amazon.ionelement.api.emptyIonStruct
+import com.amazon.ionelement.api.field
+import com.amazon.ionelement.api.ionStructOf
+
+internal class MutableStructFieldsImpl(override val fields: MutableMap<String, MutableList<AnyElement>>) :
+    MutableStructFields {
+
+    override fun get(fieldName: String): AnyElement {
+        return requireNotNull(fields[fieldName]?.firstOrNull()) {
+            "Required struct field '$fieldName' missing"
+        }
+    }
+
+    override fun getOptional(fieldName: String): AnyElement? {
+        return fields[fieldName]?.firstOrNull()
+    }
+
+    override fun getAll(fieldName: String): Iterable<AnyElement> {
+        return fields[fieldName] ?: mutableListOf()
+    }
+
+    override fun containsField(fieldName: String): Boolean {
+        return fieldName in fields
+    }
+
+    override fun set(fieldName: String, value: IonElement): MutableStructFields {
+        value as AnyElement
+        if (fieldName in fields) {
+            fields[fieldName]?.clear()
+            fields[fieldName]?.add(value)
+        } else {
+            fields[fieldName] = mutableListOf(value)
+        }
+
+        return this
+    }
+
+    override fun setAll(fields: Iterable<StructField>): MutableStructFields {
+        fields.groupByTo(mutableMapOf(), { it.name }, { it.value }).forEach { (name, values) ->
+            this.fields[name] = values
+        }
+        return this
+    }
+
+    override fun add(fieldName: String, value: IonElement): MutableStructFields {
+        value as AnyElement
+        val values = fields[fieldName]
+        if (values == null) {
+            fields[fieldName] = mutableListOf(value)
+        } else {
+            values.add(value)
+        }
+
+        return this
+    }
+
+    override fun add(field: StructField): MutableStructFields {
+        add(field.name, field.value)
+        return this
+    }
+
+    override fun remove(fieldName: String): MutableStructFields {
+        fields[fieldName]?.removeAt(0)
+        return this
+    }
+
+    override fun removeAll(fieldName: String): MutableStructFields {
+        fields[fieldName]?.clear()
+        return this
+    }
+
+    override fun plusAssign(field: StructField) {
+        add(field.name, field.value)
+    }
+
+    override fun plusAssign(fields: Iterable<StructField>) {
+        fields.forEach { add(it) }
+    }
+
+    override fun toStruct(annotations: List<String>, metas: MetaContainer): StructElement {
+        return ionStructOf(this, annotations, metas)
+    }
+
+    override fun toStruct(): StructElement {
+        return ionStructOf(this)
+    }
```

Comment:
I'm not opposed to ever adding them, but for now, there are other ways to build the struct, and I think it is better to start smaller and then add things if it does end up being an issue.

If we have `update()` or similar methods on `StructElement`, then in Java, you could write:
```java
myStruct.updateFields(f -> f.add(...)
                            .set(...));
```
And in Kotlin:
```kotlin
myStruct.update {
    add(...)
    set(...)
}
```
What do you think? Do you prefer the lambda expression (Kotlin) and consumer-builder (Java) or the traditional builder syntax?

I think this is also related to the question of whether annotations should be preserved. I think that if we have an "update" method, the name implies that it would preserve annotations and metas. If you want to get rid of the annotations and metas, you can simply create a new struct out of the `MutableStructFields` using `ionStructOf()`. (By "new struct", I'm referring to the user's intentions—you get a new instance of `StructElement` either way.)

Reply:
> What do you think? Do you prefer the lambda expression (Kotlin) and consumer-builder (Java) or the traditional builder syntax?

I prefer the former. But are you suggesting to add the `update` method in addition to `mutableStructFields()`. I like `mutableStructFields()` because you can assign it to a variable and pass it around kind of like a `StringBuilder`. With  `update` you are completing the whole process at that moment. I would propose to add both.

Reply:
I'm fine with having both.

My initial comment was that we should leave out the `toStruct()` function since there's already `ionStructOf()`. I'm not strongly opposed to it—it's just my default opinion that if something seems redundant, we should probably leave it out. If you have a strong opinion that `toStruct()` should exist, and it should be a member (as opposed to e.g. an extension function), then let's keep it.